File: 3178-167-1976-number-of-ways-to-arrive-at-destination/3178-167-1976-number-of-ways-to-arrive-at-destination.java
class Solution {
    public int countPaths(int n, int[][] roads) {
        final int mod = 1000000007;
        List<int[]>[] adj = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<>();
        }
        
        for (int[] road : roads) {
            int u = road[0], v = road[1], t = road[2];
            adj[u].add(new int[]{v, t});
            adj[v].add(new int[]{u, t});
        }
        
        long[] dist = new long[n];
        int[] ways = new int[n];
        Arrays.fill(dist, Long.MAX_VALUE);
        dist[0] = 0;
        ways[0] = 1;
        
        PriorityQueue<long[]> pq = new PriorityQueue<>((a, b) -> Long.compare(a[0], b[0]));
        pq.offer(new long[]{0, 0}); 
        
        while (!pq.isEmpty()) {
            long[] curr = pq.poll();
            long time = curr[0];
            int u = (int) curr[1];
            
            if (time > dist[u]) continue;
            
            for (int[] next : adj[u]) {
                int v = next[0], t = next[1];
                
                if (dist[v] > time + t) {
                    dist[v] = time + t;
                    pq.offer(new long[]{dist[v], v});
                    ways[v] = ways[u];
                } else if (dist[v] == time + t) {
                    ways[v] = (ways[v] + ways[u]) % mod;
                }
            }
        }
        
        return ways[n-1];
    }
}

File: 3178-167-1976-number-of-ways-to-arrive-at-destination/3178-167-1976-number-of-ways-to-arrive-at-destination.py
import heapq

class Solution:
    def countPaths(self, n: int, roads: list[list[int]]) -> int:
        mod = 10**9 + 7
        adj = [[] for _ in range(n)]
        
        for u, v, t in roads:
            adj[u].append((v, t))
            adj[v].append((u, t))
        
        dist = [float('inf')] * n
        ways = [0] * n
        dist[0] = 0
        ways[0] = 1
        
        pq = [(0, 0)]  
        
        while pq:
            time, u = heapq.heappop(pq)
            
            if time > dist[u]:
                continue
                
            for v, t in adj[u]:
                if dist[v] > time + t:
                    dist[v] = time + t
                    heapq.heappush(pq, (dist[v], v))
                    ways[v] = ways[u]
                elif dist[v] == time + t:
                    ways[v] = (ways[v] + ways[u]) % mod
        
        return ways[n-1]